tests/conftest.py

```diff
@@ -46,32 +48,70 @@ def start_mock_server():
     port = utils.free_port()
     root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
     path = os.path.join(root, "tests", "utils", "mock_server.py")
-    command = ["python", path]
+    command = [sys.executable, "-u", path]
     env = os.environ
     env["PORT"] = str(port)
     env["PYTHONPATH"] = root
-    server = subprocess.Popen(command, env=env)
+    logfile = open("tests/logs/live_mock_server.log", "w")
+    server = subprocess.Popen(command, stdout=logfile,
+                              env=env, stderr=subprocess.STDOUT, bufsize=1,
+                              close_fds=True)
     server._port = port
     server.base_url = "http://localhost:%i" % server._port
+
+    def get_ctx():
+        return requests.get(server.base_url + "/ctx").json()
+
+    def set_ctx(payload):
+        return requests.put(server.base_url + "/ctx", json=payload).json()
+
+    def reset_ctx():
+        return requests.delete(server.base_url + "/ctx").json()
+
+    server.get_ctx = get_ctx
+    server.set_ctx = set_ctx
+    server.reset_ctx = reset_ctx
+
+    started = False
     for i in range(5):
         try:
-            res = requests.get("%s/storage" % server.base_url, timeout=1)
+            res = requests.get("%s/ctx" % server.base_url, timeout=1)
             if res.status_code == 200:
+                started = True
                 break
             print("Attempting to connect but got: %s", res)
         except requests.exceptions.RequestException:
-            print("timed out")
+            print("Timed out waiting for server to start...")
+            if server.poll() is None:
+                time.sleep(1)
+            else:
+                raise ValueError("Server failed to start.")
+    if started:
+        print("Mock server listing on %s see tests/logs/live_mock_server.log" %
+              server._port)
+    else:
+        server.terminate()
+        output, _ = server.communicate(timeout=1)
+        print(output.decode("utf-8"))
+        raise ValueError("Failed to start server!  Exit code %s" % server.returncode)
     return server
 
 
-atexit.register(test_cleanup)
 start_mock_server()
+atexit.register(test_cleanup)
 
 
 @pytest.fixture
-def test_dir(runner):
-    with runner.isolated_filesystem():
-        yield runner
+def test_dir(request):
+    orig_dir = os.getcwd()
+    root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+    test_dir = os.path.join(root, "tests", "logs", request.node.name)
+    if os.path.exists(test_dir):
+        shutil.rmtree(test_dir)
+    mkdir_exists_ok(test_dir)
+    os.chdir(test_dir)
+    yield runner
+    os.chdir(orig_dir)
```

Comment:
This is fancy, any tests that use the `test_settings` or `test_dir` fixture will stage everything in `tests/logs/TEST_NAME`.  This make debugging failures much nicer as you can see the debug logs.